### ISSUE TITLE
Vulkan: artifacts with textured-object Android sample

### ISSUE THREAD

**Issue by prideout** [google/filament#749]
Shading looks bad, see blocky artifacts near the top-right of the wood shaderball in the attached Pixel 3 screenshot. A similar but more pronounced issue occurs with Mali-based devices.  Does not occur with the OpenGL backend.

![device-2019-01-28-110550](https://user-images.githubusercontent.com/1288904/51860018-6f374f00-22ed-11e9-9789-75413b02164d.png)

**Comment by prideout** [comment]
Ah, this is simply due to the lack of automatic mipmapping in the normal map.  Maybe the Vulkan backend should simply generate mipmaps via libimage and print a warning to inform devs that they should pre-generate their mips for best perf & quality.

**Comment by romainguy** [comment]
The Vulkan backend should use libimage to generate mip levels if the app invokes `generateMipmaps()`. I don't think we should print a warning since it's an opt-in behavior by the app.

**Comment by prideout** [comment]
Sounds good.

**Comment by prideout** [comment]
Actually, since generateMipmaps can be called after data has been moved to the GPU, we'll just go ahead and do this on the GPU.  After chatting with @pixelflinger we realized it is quite easy to implement using DriverAPI calls. I already have a working prototype.

**Comment by rawnsley** [comment]
This change appears to be causing rendering problems for me on Android 5.1 RK3288 devices (Mali-T760). It manifests as all the mipmap layers below the original image being black, so small triangles in the scene are painted black.

I've narrowed it down to _glBlitFramebuffer_ failing in **OpenGLDriver**::_blit()_, but I can't work out why. It works on other devices, such as an Android 9 Pixel 1, so I think it must be a quirk of the platform. If I swap _glBlitFramebuffer_ for _glClear_ the mipmap layers have color so I'm sure it's this function.

The automatic mipmap generation doesn't have this problem and I will have to temporarily revert this change until I work out what's going on.

I'm not expecting any sort of fix, but I needed somewhere to state the problem in case I gather any useful evidence in future or anyone else has the same problem.

**Comment by romainguy** [comment]
Thanks for the analysis @rawnsley. @prideout, we're probably hitting a driver bug but it could be worth taking a closer look just in case. We could fall back to `glGenerateMipmap()` on drivers with known issues?

**Comment by prideout** [comment]
Yeah adding a glGenerateMipmap fallback might be useful anyway so I'll go ahead and add that, which implies restoring the DriverAPI command that I removed.  Sigh, drivers...
